src/main/scala/openflights.scala

```diff
@@ -51,151 +51,72 @@ case class GeoRoute(route: Route, srcLat: Float, srcLng: Float, dstLat: Float, d
-        n(0).toLong,
-        n(4).replace("\"",""),
-        n(1).replace("\"",""),
-        n(6).replace("\"","")
-      )
-
-      index(code) = airline +: index(code)
-      index
-    }
+  lazy val routes: Map[RouteKey, List[Route]] = {
+    val records = loadRoutes("/data/openflights/routes.dat")
+    asMultiMap(records, route => RouteKey(route.airportSourceCode, route.airportDestCode))
   }
 
-  private def loadAirlines(path: String): List[Airline] = {
-    val records = loadCSVDataFile(path)
-    records.foldLeft(List[Airline]()) { (list, n) =>
-      list ++ List(Airline(
-        n(0).toLong,
-        n(4).replace("\"",""),
-        n(1).replace("\"",""),
-        n(6).replace("\"","")
-      ))
-    }
-  }
+  // Decoders that tell kantan.csv how to construct our case classes from the raw data
+  private implicit val airportDecoder: RowDecoder[Airport] = RowDecoder.decoder6(Airport.apply)(4, 1, 2, 3, 6, 7)
+  private implicit val airlineDecoder: RowDecoder[Airline] = RowDecoder.decoder4(Airline.apply)(0, 4, 1, 6)
+  private implicit val routeDecoder: RowDecoder[Route] = RowDecoder.decoder4(Route.apply)(0, 2, 4, 7)
 
-  private def loadAirportsAsMap(path: String): Map[String, List[Airport]] = {
-    val records = loadCSVDataFile(path)
-    val index = Map.empty[String, List[Airport]].withDefaultValue(List())
-
-    records.foldLeft(index) { (index, n) =>
-      val name = n(1).replace("\"", "")
-
-      // In case the city has a comma, it will be unintentionally split so we
-      // have one extra element in the split array.
-      val airport = if (n.length > 12) {
-        val city    = s"${n(2)}, ${n(3)}".replace("\"", "")
-        val country = n(4).replace("\"", "")
-        val code    = n(5).replace("\"", "")
-        val lat     = n(7).toFloat
-        val lng     = n(8).toFloat
-
-        Airport(code, name, city, country, lat, lng)
-      } else {
-        val city    = n(2).replace("\"", "")
-        val country = n(3).replace("\"", "")
-        val code    = n(4).replace("\"", "")
-        val lat     = n(6).toFloat
-        val lng     = n(7).toFloat
-
-        Airport(code, name, city, country, lat, lng)
-      }
-
-      val city = airport.city
-      index(city) = airport +: index(city)
-      index
-    }
+  /**
+   * Load data from a CSV file, converting records to a [[List]] of the
+   * parameterized type.
+   *
+   * Note: loads the entire file into memory eagerly; careful with huge files.
+   *
+   * @param path The path of the CSV file, as a resource on the classpath.
+   * @tparam T The class represented by records in the CSV. Requires a
+   *   [[kantan.csv.RowDecoder]] instance for the type available in scope.
+   */
+  private def loadCSVDataFile[T : RowDecoder](path: String): List[T] = {
+    val url = getClass.getResource(path)
+    // We have a static, well-formed data set -- unsafe causes no errors.
+    url.unsafeReadCsv[List, T](',', false)  // comma separator, no header
   }
```

Comment:
The syntax of this method declaration warrants some explanation regarding the complex magic of implicits for typeclasses:

```scala
def loadCSVDataFile[T : RowDecoder](path: String): List[T]
```

The `[T : RowDecoder]` part is called a [context bound]—they're also discussed, perhaps more approachably, in the Neophyte's Guide to Scala article that I linked above. This is basically syntactic sugar for this:

```scala
def loadCSVDataFile[T](path: String)(implicit ev: RowDecoder[T]): List[T]
```

This means that the function requires *evidence* to be provided within the lexical scope (an implicit) telling it how to make a `RowDecoder` of whatever type `T` is. In this case that means we have to provide a typeclass instance of `RowDecoder` for `T`. That's what we're doing with the lines like:

```scala
private implicit val airlineDecoder: RowDecoder[Airline] = RowDecoder.decoder4(Airline.apply)(0, 4, 1, 6)
```

You'll find that if you try to remove the `: RowDecoder` from `[T : RowDecoder]`, it won't compile.

[context bound]: http://docs.scala-lang.org/tutorials/FAQ/context-and-view-bounds.html